hail/src/main/scala/is/hail/expr/ir/ComputeUsesAndDefs.scala

```diff
@@ -0,0 +1,52 @@
+package is.hail.expr.ir
+
+import scala.collection.mutable
+
+case class UsesAndDefs(uses: Memo[mutable.Set[RefEquality[Ref]]], defs: Memo[RefEquality[IR]])
+
+object ComputeUsesAndDefs {
+  def apply(ir0: IR): UsesAndDefs = {
+    val uses = Memo.empty[mutable.Set[RefEquality[Ref]]]
+    val defs = Memo.empty[RefEquality[IR]]
+
+    val inAgg = InAgg(ir0)
+
+    def memoize(ir1: IR, env: Env[RefEquality[IR]], aggEnv: Env[RefEquality[IR]]) {
+      ir1 match {
+        case r@Ref(name, _) =>
+          (if (inAgg.lookup(r)) aggEnv.lookupOption(name) else env.lookupOption(name)).foreach { decl =>
+            val re = RefEquality(r)
+            uses.lookup(decl) += re
+            defs.bind(re, decl)
+          }
+        case ir =>
+          val bindings = Bindings(ir)
+          val aggBindings = AggBindings(ir)
+
+          if (bindings.nonEmpty || aggBindings.nonEmpty)
+              uses.bind(ir, mutable.Set.empty[RefEquality[Ref]])
+          Children(ir).iterator.zipWithIndex.foreach {
+            case (child: IR, idx) =>
+              var env_ = env
+              var aggEnv_ = aggEnv
+
+              bindings.foreach { binding =>
+                if (Binds(ir, binding, idx)) {
```

Comment:
It feels like our primitives are not quite right, since you're getting the bindings and then asking if each one binds.  Also, it would be nice if Bindings and Binds derived from one method so we only had to modify one thing when adding new IR that binds.

What about {Agg}Bindings(IR, Int): Array[String], which returns the set of bindings for the given child?

Reply:
That sounds like it'll work better. Can I do that in a followup PR?